dataflow/src/ops/distinct.rs

```diff
@@ -0,0 +1,252 @@
+            group_by: group_by,
+            seen: HashMap::new(),
+        }
+    }
+}
+
+impl Ingredient for Distinct {
+    /// Returns a clone of the node
+    fn take(&mut self) -> NodeOperator {
+        Clone::clone(self).into()
+    }
+
+    fn ancestors(&self) -> Vec<NodeIndex> {
+        // I think this is right?
+        vec![self.src.as_global()]
+    }
+
+    fn on_input(
+        &mut self,
+        from: LocalNodeIndex,
+        rs: Records,
+        _: &mut Tracer,
+        _: Option<&[usize]>,
+        _: &DomainNodes,
+        state: &StateMap,
+    ) -> ProcessingResult {
+        debug_assert_eq!(from, *self.src);
+
+        // Check that the records aren't empty
+        if rs.is_empty() {
+            return ProcessingResult {
+                results: rs,
+                misses: vec![],
+            };
+        }
+
+        let us = self.us.unwrap();
+        let db = state
+            .get(&*us)
+            .expect("Distinct must have its own state initialized");
+
+        let mut seen = HashMap::new();
+        let mut cleaned = Vec::new();
+
+        // Preprocess current batch to remove duplicates
+        // since those won't be caught by checking state below.
+        for rec in rs.iter() {
+            let group = rec.iter()
+                .enumerate()
+                .filter_map(|(i, v)| {
+                    if self.group_by.iter().any(|col| col == &i) {
+                        Some(v)
+                    } else {
+                        None
+                    }
+                })
+                .cloned()
+                .collect::<Vec<_>>();
+
+            let positive = rec.is_positive();
+            if seen.contains_key(&group){
+                if !positive{
+                    seen.remove(&group);
+                    cleaned.push(rec.clone());
+                }
+            } else {
+                cleaned.push(rec.clone());
+                seen.insert(group.clone(), true);
+            }
+        }
+
+        let mut misses = Vec::new();
+        let mut output = Vec::new();
+        for rec in cleaned.iter() {
+            let group_by = &self.group_by[..];
+            let group = rec.iter()
+                .enumerate()
+                .filter_map(|(i, v)| {
+                    if self.group_by.iter().any(|col| col == &i) {
+                        Some(v)
+                    } else {
+                        None
+                    }
+                })
+                .cloned()
+                .collect::<Vec<_>>();
+            let positive = rec.is_positive();
+            if positive {
+                match db.lookup(group_by, &KeyType::from(&group[..])) {
+                    LookupResult::Some(rr) => {
+                        //println!("record {:?}", rr);
+                        if rr.len() == 0 {
+                            output.push(rec.clone());
+                        }
+                    },
+                    LookupResult::Missing => {
```

Comment:
I would put `unimplemented!("Distinct does not yet support partial")` here so we don't miss (heh) this case should it happen due to a bug.